src/nxt_http_return.c

```diff
@@ -87,27 +88,139 @@ nxt_http_return(nxt_task_t *task, nxt_http_request_t *r,
+            return NULL;
+        }
 
-        field->value = conf->location.start;
-        field->value_length = conf->location.length;
+        nxt_var_query(task, r->var_query, conf->location, &ctx->location);
+
+        nxt_var_query_resolve(task, r->var_query, ctx,
+                              nxt_http_return_send_ready,
+                              nxt_http_return_var_error);
+    }
+
+    return NULL;
+}
+
+
+static nxt_int_t
+nxt_http_return_encode(nxt_mp_t *mp, nxt_str_t *encoded,
+    const nxt_str_t *location)
+{
+    nxt_str_t   *loc;
+    nxt_uint_t  encode;
+
+    if (nxt_is_complex_uri_encoded(location->start, location->length)) {
+        loc = nxt_str_dup(mp, encoded, location);
+        if (nxt_slow_path(loc == NULL)) {
+            return NXT_ERROR;
+        }
+
+        return NXT_OK;
+    }
+
+    encode = nxt_encode_complex_uri(NULL, location->start, location->length);
+    encoded->length = location->length + encode * 2;
+
+    encoded->start = nxt_mp_nget(mp, encoded->length);
+    if (nxt_slow_path(encoded->start == NULL)) {
+        return NXT_ERROR;
     }
 
+    nxt_encode_complex_uri(encoded->start, location->start, location->length);
+
+    return NXT_OK;
+}
+
+
+static void
+nxt_http_return_send_ready(nxt_task_t *task, void *obj, void *data)
+{
+    nxt_int_t               ret;
+    nxt_str_t               *loc;
+    nxt_http_field_t        *field;
+    nxt_http_action_t       *action;
+    nxt_http_request_t      *r;
+    nxt_http_return_ctx_t   *ctx;
+    nxt_http_return_conf_t  *conf;
+
+    r = obj;
+    ctx = data;
+    action = ctx->action;
+    conf = action->u.conf;
+
+    if ((conf->loc_is_const && conf->encoded.length == 0)
+        || (!conf->loc_is_const && ctx->location.length == 0))
+    {
+        goto send_state;
+    }
+
+    if (conf->loc_is_const) {
+        loc = nxt_str_dup(r->mem_pool, &ctx->encoded, &conf->encoded);
+        if (nxt_slow_path(loc == NULL)) {
+            nxt_http_request_error(task, r, NXT_HTTP_INTERNAL_SERVER_ERROR);
+            return;
+        }
+
+    } else {
+        ret = nxt_http_return_encode(r->mem_pool, &ctx->encoded,
+                                     &ctx->location);
+        if (nxt_slow_path(ret == NXT_ERROR)) {
+            nxt_http_request_error(task, r, NXT_HTTP_INTERNAL_SERVER_ERROR);
+            return;
+        }
+    }
+
+    field = nxt_list_zero_add(r->resp.fields);
+    if (nxt_slow_path(field == NULL)) {
+        nxt_http_request_error(task, r, NXT_HTTP_INTERNAL_SERVER_ERROR);
+        return;
+    }
+
+    nxt_http_field_name_set(field, "Location");
+
+    field->value = ctx->encoded.start;
+    field->value_length = ctx->encoded.length;
+
+send_state:
     r->state = &nxt_http_return_send_state;
 
     nxt_http_request_header_send(task, r, NULL, NULL);
 
-    return NULL;
+    return;
+}
+
+
+static void
+nxt_http_return_var_error(nxt_task_t *task, void *obj, void *data)
```

Comment:
I suggest renaming `nxt_http_return_send_error` instead of `nxt_http_return_var_error`.
I also need to rename similar function in other files in my related patches.

Reply:
I prefer to keep it as `_var_error()` for now for consistency, as is used in other parts of the code (`"chroot"` and `"share"`), and then we can add a separate commit to rename them all at once.  What do you think?

Reply:
Agree, it makes sense.